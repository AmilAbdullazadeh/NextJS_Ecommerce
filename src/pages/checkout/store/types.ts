export interface UpdateBillingFullNameActionPayload {
    fullName: string;
}

export interface UpdateBillingAddressActionPayload {
    address: string;
}

export interface UpdateCardNumberActionPayload {
    number: string;
}

export interface UpdateCardNameActionPayload {
    name: string;
}

export interface UpdateCardExpiryActionPayload {
    expiry: string;
}

export interface UpdateCardCvcActionPayload {
    cvc: string;
}
