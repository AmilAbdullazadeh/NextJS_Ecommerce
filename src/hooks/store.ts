import { TypedUseSelectorHook, useSelector, useDispatch } from 'react-redux';
import { RootState, AppDispatch } from '../store/store';

export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const useAppDispatch = () => useDispatch<AppDispatch>();
