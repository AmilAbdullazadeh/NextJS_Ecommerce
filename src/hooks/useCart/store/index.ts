export { default as cartReducer } from './reducer';
