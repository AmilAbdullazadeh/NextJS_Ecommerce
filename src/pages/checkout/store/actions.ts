import { createAction } from '@reduxjs/toolkit';

import { withPrefix } from '@utils';

import {
    UpdateBillingFullNameActionPayload,
    UpdateBillingAddressActionPayload,
    UpdateCardNumberActionPayload,
    UpdateCardNameActionPayload,
    UpdateCardExpiryActionPayload,
    UpdateCardCvcActionPayload,
} from './types';

const nameSpace = 'CHECKOUT';
const withNameSpacePrefix = withPrefix(nameSpace);

export const updateBillingFullNameAction =
    createAction<UpdateBillingFullNameActionPayload>(
        withNameSpacePrefix('update_billing_name')
    );

export const updateBillingAddressAction =
    createAction<UpdateBillingAddressActionPayload>(
        withNameSpacePrefix('update_billing_address')
    );

export const updateCardNumberAction =
    createAction<UpdateCardNumberActionPayload>(
        withNameSpacePrefix('update_card_number')
    );

export const updateCardNameAction = createAction<UpdateCardNameActionPayload>(
    withNameSpacePrefix('update_card_name')
);

export const updateCardExpiryAction =
    createAction<UpdateCardExpiryActionPayload>(
        withNameSpacePrefix('update_card_expiry')
    );

export const updateCardCvcAction = createAction<UpdateCardCvcActionPayload>(
    withNameSpacePrefix('update_card_cvc')
);

export const clearCheckoutInfoAction = createAction(
    withNameSpacePrefix('clear')
);
