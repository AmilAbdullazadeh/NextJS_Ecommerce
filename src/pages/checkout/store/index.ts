export { default as checkoutReducer } from './reducer';
