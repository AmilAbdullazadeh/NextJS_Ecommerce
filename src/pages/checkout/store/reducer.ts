import { createReducer } from '@reduxjs/toolkit';

import type { ActionHandler } from '@core';

import {
    updateBillingFullNameAction,
    updateBillingAddressAction,
    updateCardNumberAction,
    updateCardNameAction,
    updateCardExpiryAction,
    updateCardCvcAction,
    clearCheckoutInfoAction,
} from './actions';
import {
    UpdateBillingFullNameActionPayload,
    UpdateBillingAddressActionPayload,
    UpdateCardNumberActionPayload,
    UpdateCardNameActionPayload,
    UpdateCardExpiryActionPayload,
    UpdateCardCvcActionPayload,
} from './types';

export interface CheckoutState {
    billingInfo: {
        fullName: string;
        address: string;
    };
    paymentInfo: {
        /**
         * the card property could be ommited for our project since it is the only
         * supported payment method, but in a professional project the paymentInfo
         * would have more options.
         */
        card: {
            number: string;
            name: string;
            expiry: string;
            cvc: string;
        };
    };
}

export const initialState: CheckoutState = {
    billingInfo: {
        fullName: '',
        address: '',
    },
    paymentInfo: {
        card: {
            number: '',
            name: '',
            expiry: '',
            cvc: '',
        },
    },
};

const handleUpdateBillingFullName: ActionHandler<
    CheckoutState,
    UpdateBillingFullNameActionPayload
> = (state, { payload: { fullName } }) => {
    state.billingInfo.fullName = fullName;
};

const handleUpdateBillingAddress: ActionHandler<
    CheckoutState,
    UpdateBillingAddressActionPayload
> = (state, { payload: { address } }) => {
    state.billingInfo.address = address;
};

const handleUpdateCardNumber: ActionHandler<
    CheckoutState,
    UpdateCardNumberActionPayload
> = (state, { payload: { number } }) => {
    state.paymentInfo.card.number = number;
};

const handleUpdateCardName: ActionHandler<
    CheckoutState,
    UpdateCardNameActionPayload
> = (state, { payload: { name } }) => {
    state.paymentInfo.card.name = name;
};

const handleUpdateCardExpiryName: ActionHandler<
    CheckoutState,
    UpdateCardExpiryActionPayload
> = (state, { payload: { expiry } }) => {
    state.paymentInfo.card.expiry = expiry;
};

const handleUpdateCardCvcName: ActionHandler<
    CheckoutState,
    UpdateCardCvcActionPayload
> = (state, { payload: { cvc } }) => {
    state.paymentInfo.card.cvc = cvc;
};

const checkoutReducer = createReducer(initialState, {
    [updateBillingFullNameAction.type]: handleUpdateBillingFullName,
    [updateBillingAddressAction.type]: handleUpdateBillingAddress,
    [updateCardNumberAction.type]: handleUpdateCardNumber,
    [updateCardNameAction.type]: handleUpdateCardName,
    [updateCardExpiryAction.type]: handleUpdateCardExpiryName,
    [updateCardCvcAction.type]: handleUpdateCardCvcName,
    [clearCheckoutInfoAction.type]: () => initialState,
});

export default checkoutReducer;
