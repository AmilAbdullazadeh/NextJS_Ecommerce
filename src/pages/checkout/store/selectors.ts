import { createSelector } from '@reduxjs/toolkit';

import { RootState } from '@store/store';

export const cardNumberSelector: (state: RootState) => string = state =>
    state.checkout.paymentInfo.card.number;

export const cardNameSelector: (state: RootState) => string = state =>
    state.checkout.paymentInfo.card.name;

export const cardExpirySelector: (state: RootState) => string = state =>
    state.checkout.paymentInfo.card.expiry;

export const cardCvcSelector: (state: RootState) => string = state =>
    state.checkout.paymentInfo.card.cvc;

export const isValidCardNumberSelector: (state: RootState) => boolean =
    createSelector(cardNumberSelector, cardNumber => {
        const containsOnlyNumbers = Array.from(cardNumber).every(number =>
            number.match(/[0-9]/)
        );
        const expectedLength = 16;
        const hasCorrectLength = cardNumber.length === expectedLength;

        return containsOnlyNumbers && hasCorrectLength;
    });

export const isValidCardNameSelector: (state: RootState) => boolean =
    createSelector(cardNameSelector, cardName => {
        const containsValidChars = Array.from(cardName).every(letter =>
            letter.match(/[a-z A-Z]/)
        );

        const maxLength = 60;
        const isLessThanMaxLength =
            cardName.length > 0 && cardName.length <= maxLength;

        return containsValidChars && isLessThanMaxLength;
    });

export const isValidCardExpirySelector: (state: RootState) => boolean =
    createSelector(cardExpirySelector, expiry => {
        const expectedLength = 4;

        const monthIndexStart = 0;
        const monthIndexEnd = 2;
        const yearIndexStart = 2;
        const yearIndexEnd = 4;
        const lastTwoDigits = -2;

        const monthField = Number(
            expiry.substring(monthIndexStart, monthIndexEnd)
        );

        const monthsInYear = 12;

        const isCorrectMonthField =
            monthField > 0 && monthField <= monthsInYear;

        const yearField = Number(
            expiry.substring(yearIndexStart, yearIndexEnd)
        );

        const isCorrectYearField =
            yearField >=
            Number(new Date().getFullYear().toString().substr(lastTwoDigits));

        const hasCorrectLength = expiry.length === expectedLength;

        return hasCorrectLength && isCorrectMonthField && isCorrectYearField;
    });

export const isValidCardCvcSelector: (state: RootState) => boolean =
    createSelector(cardCvcSelector, cvc => {
        const expectedLength = 3;
        const containsOnlyNumbers = Array.from(cvc).every(number =>
            number.match(/[0-9]/)
        );

        const hasCorrectLength = cvc.length === expectedLength;

        return containsOnlyNumbers && hasCorrectLength;
    });

export const isValidCardPaymentFormSelector: (state: RootState) => boolean =
    createSelector(
        isValidCardNumberSelector,
        isValidCardNameSelector,
        isValidCardExpirySelector,
        isValidCardCvcSelector,
        (isValidCardNumber, isValidCardName, isValidExpiry, isValidCvc) =>
            isValidCardNumber && isValidCardName && isValidExpiry && isValidCvc
    );

export const billingFullNameSelector: (state: RootState) => string = state =>
    state.checkout.billingInfo.fullName;

export const billingAddressSelector: (state: RootState) => string = state =>
    state.checkout.billingInfo.address;

export const isValidBillingFullNameSelector: (state: RootState) => boolean =
    createSelector(billingFullNameSelector, name => {
        const containsValidChars = Array.from(name).every(char =>
            char.match(/[a-z A-Z]/)
        );
        const maxLength = 60;
        const hasCorrectLength = name.length > 0 && name.length <= maxLength;

        return containsValidChars && hasCorrectLength;
    });

export const isValidBillingAddressSelector: (state: RootState) => boolean =
    createSelector(billingAddressSelector, address => {
        const maxLength = 30;

        const hasCorrectLength =
            address.length > 0 && address.length <= maxLength;

        return hasCorrectLength;
    });

export const isValidBillingInfoForm: (state: RootState) => boolean =
    createSelector(
        isValidBillingFullNameSelector,
        isValidBillingAddressSelector,
        (isValidBillingFullName, isValidBillingAddress) =>
            isValidBillingFullName && isValidBillingAddress
    );
