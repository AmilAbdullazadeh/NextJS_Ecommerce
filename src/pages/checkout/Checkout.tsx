import React, { useEffect, useState } from 'react';
import { useRouter } from 'next/router';
import Box from '@material-ui/core/Box';
import Stepper from '@material-ui/core/Stepper';
import Step from '@material-ui/core/Step';
import StepContent from '@material-ui/core/StepContent';
import StepButton from '@material-ui/core/StepButton';
import Button from '@material-ui/core/Button';
import CircularProgress from '@material-ui/core/CircularProgress';

import { CardPaymentForm, BillingInformationForm } from './components';

import {
    useCart,
    useAppSelector as useSelect,
    useAppDispatch,
    emptyCartAction,
} from '@hooks';
import { CartProduct, landingPage } from '@core';

import {
    isValidBillingInfoForm,
    isValidCardPaymentFormSelector,
} from './store/selectors';
import { clearCheckoutInfoAction } from './store/actions';

const billingInfoStep = 0;
const paymentMethodStep = 1;
const finishStep = 2;

const timeout = 3000;

export const handleCheckout: (items: CartProduct[]) => Promise<string> = () =>
    new Promise(resolve => {
        setTimeout(() => {
            resolve('order completed');
        }, timeout);
    });

const getStepContent: (
    step: number,
    isPendingCheckout: boolean
) => React.ReactNode = (step, isPendingCheckout) => {
    switch (step) {
        case billingInfoStep:
            return <BillingInformationForm />;
        case paymentMethodStep:
            return <CardPaymentForm />;
        case finishStep:
            return (
                <div>
                    {isPendingCheckout ? (
                        <CircularProgress data-testid="loading-spinner" />
                    ) : (
                        'complete order'
                    )}
                </div>
            );
        default:
            'unexpected state';
    }
};

const Checkout: React.FC = () => {
    const router = useRouter();
    const dispatch = useAppDispatch();
    const { products: items } = useCart();
    const [activeStep, setActiveStep] = useState(0);
    const isValidBillingInfoStep = useSelect(isValidBillingInfoForm);
    const isValidCardPaymentStep = useSelect(isValidCardPaymentFormSelector);
    const [isPendingCheckout, setIsPendingCheckout] = useState(false);

    const cleanUpCheckoutInformation = () =>
        dispatch(clearCheckoutInfoAction());

    const cleanUpCart = () => dispatch(emptyCartAction());

    const completeOrder = async () => {
        const result = await handleCheckout(items);
        console.log(result);
        cleanUpCheckoutInformation();
        cleanUpCart();

        router.push(landingPage());
    };

    const steps = ['Billing Address', 'Payment Method', 'Finish'];

    useEffect(() => {
        window.addEventListener('beforeunload', cleanUpCheckoutInformation);

        return () => {
            cleanUpCheckoutInformation();
            window.removeEventListener(
                'beforeunload',
                cleanUpCheckoutInformation
            );
        };
    }, []);

    useEffect(() => {
        if (items.length === 0) {
            router.push(landingPage());
        }
    });

    const handleStep = (step: number) => {
        setActiveStep(step);
    };

    const handleNextStep = () => {
        if (activeStep === finishStep) {
            setIsPendingCheckout(true);
            completeOrder();
        } else {
            setActiveStep(prevActiveStep => prevActiveStep + 1);
        }
    };

    const handleBackStep = () => {
        setActiveStep(prevActiveStep => prevActiveStep - 1);
    };

    const isNextStepBtnDisabled = (step: number) => {
        switch (step) {
            case billingInfoStep:
                return !isValidBillingInfoStep;
            case paymentMethodStep:
                return !isValidCardPaymentStep;
            case finishStep:
                return false;
            default:
                true;
        }
    };

    return (
        <Box>
            <Stepper
                activeStep={activeStep}
                orientation="vertical"
                style={{ width: '80vw' }}
            >
                {steps.map((label, index) => (
                    <Step key={label}>
                        <StepButton onClick={() => handleStep(index)}>
                            {label}
                        </StepButton>
                        <StepContent>
                            <Box minHeight={0}>
                                {getStepContent(activeStep, isPendingCheckout)}
                            </Box>

                            <Box mt={2}>
                                <Button
                                    disabled={activeStep === 0}
                                    aria-label="previous-step"
                                    onClick={handleBackStep}
                                >
                                    Back
                                </Button>
                                <Button
                                    variant="text"
                                    aria-label="next-step"
                                    onClick={handleNextStep}
                                    style={{
                                        backgroundColor: '#FFA41C',
                                    }}
                                    disabled={isNextStepBtnDisabled(activeStep)}
                                >
                                    {activeStep === steps.length - 1
                                        ? 'Finish'
                                        : 'Next'}
                                </Button>
                            </Box>
                        </StepContent>
                    </Step>
                ))}
            </Stepper>
        </Box>
    );
};

export default Checkout;
