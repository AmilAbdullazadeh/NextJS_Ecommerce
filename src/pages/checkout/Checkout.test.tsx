import React from 'react';

import { mockState, renderWithProviders, act } from '@testUtils';
import { RootState } from '@store/store';
import { checkoutPage, landingPage } from '@core';

import Checkout from './Checkout';

describe('Checkout page', () => {
    afterAll(() => {
        jest.restoreAllMocks();
    });
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const useRouter = jest.spyOn(require('next/router'), 'useRouter');
    const mockPush = jest.fn();
    useRouter.mockImplementation(() => ({
        pathname: checkoutPage(),
        push: mockPush,
    }));
    const initialState: RootState = { ...mockState };
    it('renders correctly', () => {
        const {
            container: { firstChild },
        } = renderWithProviders(<Checkout />, { initialState });

        expect(firstChild).toMatchSnapshot();
    });

    it('redirects to landing page if cart has no items', () => {
        const noCartItemsState: RootState = {
            ...mockState,
            cart: {
                ...mockState.cart,
                products: [],
            },
        };

        renderWithProviders(<Checkout />, { initialState: noCartItemsState });

        expect(mockPush).toHaveBeenCalledWith(landingPage());
    });

    describe('switches between checkout steps', () => {
        it('switches between steps', () => {
            const {
                getAllByText,
                getByLabelText,
                queryAllByTestId,
                getAllByRole,
            } = renderWithProviders(<Checkout />, {
                initialState,
            });

            const previousBtn = getByLabelText('previous-step');
            const nextBtn = getByLabelText('next-step');

            expect(previousBtn).toBeDisabled();

            nextBtn.click();

            expect(
                queryAllByTestId('payment-info-form').length
            ).toBeGreaterThan(0);

            nextBtn.click();

            expect(getAllByText('complete order').length).toBeGreaterThan(0);

            previousBtn.click();

            expect(
                queryAllByTestId('payment-info-form').length
            ).toBeGreaterThan(0);

            previousBtn.click();

            expect(
                getAllByRole('textbox', { name: /address/i }).length
            ).toBeGreaterThan(0);
        });

        it('switches to a different step by clicking on it', () => {
            const { getByText, getByLabelText, getAllByRole } =
                renderWithProviders(<Checkout />, { initialState });

            const nextBtn = getByLabelText('next-step');

            nextBtn.click();

            const previousStepLabel = getByText(/billing address/i);

            previousStepLabel.click();

            expect(
                getAllByRole('textbox', {
                    name: /full name/i,
                }).length
            ).toBeGreaterThan(0);
        });

        it('shows spinner while processing order', () => {
            jest.useFakeTimers();

            const { getByLabelText, getAllByTestId } = renderWithProviders(
                <Checkout />,
                { initialState }
            );

            const nextBtn = getByLabelText('next-step');

            act(() => {
                nextBtn.click(); // proceed to payment info
            });
            act(() => {
                nextBtn.click(); // proceed to last step
            });
            act(() => {
                nextBtn.click(); // finish order
            });

            act(() => {
                jest.runAllTimers();
            });

            expect(getAllByTestId('loading-spinner').length).toBeGreaterThan(0);

            jest.useRealTimers();
        });
    });
});
