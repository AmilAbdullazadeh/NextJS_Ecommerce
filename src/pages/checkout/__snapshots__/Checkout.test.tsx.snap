// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkout page renders correctly 1`] = `
<div
  class="MuiBox-root MuiBox-root-1"
>
  <div
    class="MuiPaper-root MuiStepper-root MuiStepper-vertical MuiPaper-elevation0"
    style="width: 80vw;"
  >
    <div
      class="MuiStep-root MuiStep-vertical"
    >
      <button
        class="MuiButtonBase-root MuiStepButton-root MuiStepButton-vertical"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiStepLabel-root MuiStepLabel-vertical"
        >
          <span
            class="MuiStepLabel-iconContainer"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiStepIcon-root MuiStepIcon-active"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <circle
                cx="12"
                cy="12"
                r="12"
              />
              <text
                class="MuiStepIcon-text"
                text-anchor="middle"
                x="12"
                y="16"
              >
                1
              </text>
            </svg>
          </span>
          <span
            class="MuiStepLabel-labelContainer"
          >
            <span
              class="MuiTypography-root MuiStepLabel-label MuiStepLabel-active MuiTypography-body2 MuiTypography-displayBlock"
            >
              Billing Address
            </span>
          </span>
        </span>
        <span
          class="MuiTouchRipple-root MuiStepButton-touchRipple"
        />
      </button>
      <div
        class="MuiStepContent-root"
        icon="1"
      >
        <div
          class="MuiCollapse-root MuiStepContent-transition MuiCollapse-entered"
          style="min-height: 0px;"
        >
          <div
            class="MuiCollapse-wrapper"
          >
            <div
              class="MuiCollapse-wrapperInner"
            >
              <div
                class="MuiBox-root MuiBox-root-2"
              >
                <div
                  id="billing-info-form"
                >
                  <div
                    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-content-xs-center"
                  >
                    <div
                      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
                    >
                      <div
                        class="MuiFormControl-root MuiTextField-root"
                      >
                        <label
                          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
                          data-shrink="true"
                        >
                          full name
                        </label>
                        <div
                          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                        >
                          <input
                            aria-invalid="false"
                            aria-label="full name"
                            class="MuiInputBase-input MuiOutlinedInput-input"
                            data-testid="full name"
                            type="text"
                            value="Jason Smith"
                          />
                          <fieldset
                            aria-hidden="true"
                            class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
                          >
                            <legend
                              class="PrivateNotchedOutline-legendLabelled-5 PrivateNotchedOutline-legendNotched-6"
                            >
                              <span>
                                full name
                              </span>
                            </legend>
                          </fieldset>
                        </div>
                        
                      </div>
                    </div>
                    <div
                      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
                    >
                      <div
                        class="MuiFormControl-root MuiTextField-root"
                      >
                        <label
                          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
                          data-shrink="true"
                        >
                          address
                        </label>
                        <div
                          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                        >
                          <input
                            aria-invalid="false"
                            aria-label="address"
                            class="MuiInputBase-input MuiOutlinedInput-input"
                            data-testid="address"
                            type="text"
                            value="Black Mesa 13"
                          />
                          <fieldset
                            aria-hidden="true"
                            class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
                          >
                            <legend
                              class="PrivateNotchedOutline-legendLabelled-5 PrivateNotchedOutline-legendNotched-6"
                            >
                              <span>
                                address
                              </span>
                            </legend>
                          </fieldset>
                        </div>
                        
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="MuiBox-root MuiBox-root-7"
              >
                <button
                  aria-label="previous-step"
                  class="MuiButtonBase-root MuiButton-root MuiButton-text Mui-disabled Mui-disabled"
                  disabled=""
                  tabindex="-1"
                  type="button"
                >
                  <span
                    class="MuiButton-label"
                  >
                    Back
                  </span>
                </button>
                <button
                  aria-label="next-step"
                  class="MuiButtonBase-root MuiButton-root MuiButton-text"
                  style="background-color: rgb(255, 164, 28);"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiButton-label"
                  >
                    Next
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiStepConnector-root MuiStepConnector-vertical Mui-disabled"
    >
      <span
        class="MuiStepConnector-line MuiStepConnector-lineVertical"
      />
    </div>
    <div
      class="MuiStep-root MuiStep-vertical"
    >
      <button
        class="MuiButtonBase-root MuiStepButton-root MuiStepButton-vertical Mui-disabled"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiStepLabel-root MuiStepLabel-vertical Mui-disabled"
        >
          <span
            class="MuiStepLabel-iconContainer"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiStepIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <circle
                cx="12"
                cy="12"
                r="12"
              />
              <text
                class="MuiStepIcon-text"
                text-anchor="middle"
                x="12"
                y="16"
              >
                2
              </text>
            </svg>
          </span>
          <span
            class="MuiStepLabel-labelContainer"
          >
            <span
              class="MuiTypography-root MuiStepLabel-label MuiTypography-body2 MuiTypography-displayBlock"
            >
              Payment Method
            </span>
          </span>
        </span>
      </button>
      <div
        class="MuiStepContent-root"
        disabled=""
        icon="2"
      />
    </div>
    <div
      class="MuiStepConnector-root MuiStepConnector-vertical Mui-disabled"
    >
      <span
        class="MuiStepConnector-line MuiStepConnector-lineVertical"
      />
    </div>
    <div
      class="MuiStep-root MuiStep-vertical"
    >
      <button
        class="MuiButtonBase-root MuiStepButton-root MuiStepButton-vertical Mui-disabled"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiStepLabel-root MuiStepLabel-vertical Mui-disabled"
        >
          <span
            class="MuiStepLabel-iconContainer"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiStepIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <circle
                cx="12"
                cy="12"
                r="12"
              />
              <text
                class="MuiStepIcon-text"
                text-anchor="middle"
                x="12"
                y="16"
              >
                3
              </text>
            </svg>
          </span>
          <span
            class="MuiStepLabel-labelContainer"
          >
            <span
              class="MuiTypography-root MuiStepLabel-label MuiTypography-body2 MuiTypography-displayBlock"
            >
              Finish
            </span>
          </span>
        </span>
      </button>
      <div
        class="MuiStepContent-root MuiStepContent-last"
        disabled=""
        icon="3"
      />
    </div>
  </div>
</div>
`;
